BEIMA.Backend/ReportService/ReportWriter.cs

```diff
@@ -0,0 +1,286 @@
+                            csvWriter.Flush();
+                            dataStream.Position = 0;
+                            dataStream.CopyTo(fileInZipStream);
+                        }
+                    }
+                }
+                zipFileBytes = outStream.ToArray();
+            }
+            return zipFileBytes;
+        }
+
+        /// <summary>
+        /// Creates a list of strings filled with all of the property names in the Device object. The Fields
+        /// property is replaced with the field names contained in the parameter deviceType's Fields dictionary
+        /// </summary>
+        /// <param name="deviceType"></param>
+        /// <param name="delimiter"></param>
+        /// <returns></returns>
+        private static string GenerateDeviceHeaders(DeviceType deviceType, string delimiter = ",")
+        {
+            var headers = new List<string>();
+            foreach (var deviceProp in typeof(Device).GetProperties())
+            {
+                if (deviceProp.Name == "Fields")
+                {
+                    // Add all of the DeviceType field names
+                    var fieldProps = deviceType.Fields.ToDictionary().Values.Select(field => { return (string)field; });
+                    headers.AddRange(fieldProps);
+                }
+                else if (deviceProp.Name == "Location")
+                {
+                    // Add all of the DeviceLocation prop names
+                    var locProps = typeof(DeviceLocation).GetProperties().Select(prop => prop.Name);
+                    headers.AddRange(locProps);
+                }
+                else if (deviceProp.Name == "LastModified")
+                {
+                    // Add all of the DeviceLastModified prop names
+                    var modProps = typeof(DeviceLastModified).GetProperties().Select(prop => prop.Name);
+                    headers.AddRange(modProps);
+                }
+                else if (deviceProp.Name != "Photo" && deviceProp.Name != "Files")
+                {
+                    // Add all 'primative' prop names
+                    headers.Add(deviceProp.Name);
+                }
+            }
+            return string.Join(delimiter, headers);
+        }
+
+        /// <summary>
+        /// Creates a list of strings filled with all of the property values in the parameter Device object. The Fields
+        /// property values are based on the keys in the parameter deviceType's fields and the values of the parameter device's fields.
+        /// </summary>
+        /// <param name="device"></param>
+        /// <param name="deviceType"></param>
+        /// <param name="delimiter"></param>
+        /// <returns></returns>
+        private static string GenerateDeviceValues(Device device, DeviceType deviceType, string delimiter = ",")
+        {
+            var values = new List<string>();
+            foreach (var deviceProp in typeof(Device).GetProperties())
```

Comment:
So I what you have here is probably working for now, but I think that we might not be able to guarantee that the values here are being placed in the same column as the previously generated column headers. I think it would be a good idea to also pass in the array of column headers (or the header string) to this method so that we can check to make sure that we guarantee the values in here are being placed into the correct columns, and we aren't mixing any of them up.